File: services/youtube-factory-api/app/api/v1/endpoints/analytics.py
"""Analytics and dashboard endpoints."""
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.models.project import Project
from app.models.pipeline import Pipeline
from app.models.job import Job

router = APIRouter()


@router.get("/dashboard")
async def get_dashboard_stats(
    db: Annotated[AsyncSession, Depends(get_db)],
) -> dict:
    project_status_rows = (
        await db.execute(
            select(Project.status, func.count(Project.id).label("count")).group_by(Project.status)
        )
    ).all()

    job_status_rows = (
        await db.execute(
            select(Job.status, func.count(Job.id).label("count")).group_by(Job.status)
        )
    ).all()

    projects_by_status = [{"status": r[0], "count": r[1]} for r in project_status_rows]
    jobs_by_status = {r[0]: r[1] for r in job_status_rows}

    total_projects = sum(r["count"] for r in projects_by_status)
    total_videos = next((r["count"] for r in projects_by_status if r["status"] == "completed"), 0)
    active_jobs = jobs_by_status.get("running", 0)
    queued_jobs = jobs_by_status.get("pending", 0)
    failed_jobs = jobs_by_status.get("failed", 0)
    completed_jobs = jobs_by_status.get("completed", 0)
    total_processed = completed_jobs + failed_jobs
    success_rate = round((completed_jobs / total_processed * 100), 1) if total_processed > 0 else 0

    recent_projects = (
        await db.execute(
            select(Project).order_by(Project.updated_at.desc()).limit(10)
        )
    ).scalars().all()

    recent_activity = [
        {
            "id": p.id,
            "type": "project",
            "message": f'Project "{p.title}" — {p.status}',
            "projectId": p.id,
            "status": p.status,
            "timestamp": p.updated_at.isoformat(),
        }
        for p in recent_projects
    ]

    return {
        "totalProjects": total_projects,
        "totalVideos": total_videos,
        "activeJobs": active_jobs,
        "queuedJobs": queued_jobs,
        "failedJobs": failed_jobs,
        "successRate": success_rate,
        "totalRuntime": 0,
        "projectsByStatus": projects_by_status,
        "recentActivity": recent_activity,
    }


@router.get("/pipeline-activity")
async def get_pipeline_activity(
    db: Annotated[AsyncSession, Depends(get_db)],
    limit: int = Query(20, le=100),
) -> dict:
    pipelines = (
        await db.execute(select(Pipeline).order_by(Pipeline.started_at.desc()).limit(limit))
    ).scalars().all()

    items = [
        {
            "id": p.id,
            "type": "pipeline",
            "message": f"Pipeline {p.status}" + (f" — stage: {p.current_stage}" if p.current_stage else ""),
            "projectId": p.project_id,
            "status": p.status,
            "timestamp": p.started_at.isoformat(),
        }
        for p in pipelines
    ]
    return {"items": items}


@router.get("/stage-breakdown")
async def get_stage_breakdown(
    db: Annotated[AsyncSession, Depends(get_db)],
) -> dict:
    stage_names = [
        "research", "script", "scene_planning", "image_generation",
        "voice_generation", "video_editing", "subtitle_generation",
        "thumbnail_generation", "seo_generation", "upload",
    ]
    return {
        "stages": [
            {"name": name, "completed": 0, "failed": 0, "avgDurationMs": 0}
            for name in stage_names
        ]
    }

File: artifacts/api-server/src/routes/analytics.ts
import { Router } from "express";
import { db } from "@workspace/db";
import { projectsTable, pipelinesTable, jobsTable } from "@workspace/db";
import { sql, desc } from "drizzle-orm";

const router = Router();

// GET /analytics/dashboard
router.get("/dashboard", async (req, res) => {
  try {
    const [
      projectStats,
      jobStats,
      recentProjects,
    ] = await Promise.all([
      db
        .select({
          status: projectsTable.status,
          count: sql<number>`cast(count(*) as integer)`,
        })
        .from(projectsTable)
        .groupBy(projectsTable.status),
      db
        .select({
          status: jobsTable.status,
          count: sql<number>`cast(count(*) as integer)`,
        })
        .from(jobsTable)
        .groupBy(jobsTable.status),
      db
        .select()
        .from(projectsTable)
        .orderBy(desc(projectsTable.updatedAt))
        .limit(10),
    ]);

    const totalProjects = projectStats.reduce((sum, r) => sum + r.count, 0);
    const totalVideos = projectStats
      .filter((r) => r.status === "completed")
      .reduce((sum, r) => sum + r.count, 0);
    const activeJobs = jobStats
      .filter((r) => r.status === "running")
      .reduce((sum, r) => sum + r.count, 0);
    const queuedJobs = jobStats
      .filter((r) => r.status === "pending")
      .reduce((sum, r) => sum + r.count, 0);
    const failedJobs = jobStats
      .filter((r) => r.status === "failed")
      .reduce((sum, r) => sum + r.count, 0);
    const completedJobs = jobStats
      .filter((r) => r.status === "completed")
      .reduce((sum, r) => sum + r.count, 0);
    const totalJobsProcessed = completedJobs + failedJobs;
    const successRate =
      totalJobsProcessed > 0
        ? Math.round((completedJobs / totalJobsProcessed) * 100)
        : 0;

    const recentActivity = recentProjects.map((p) => ({
      id: p.id,
      type: "project",
      message: `Project "${p.title}" — ${p.status}`,
      projectId: p.id,
      status: p.status,
      timestamp: p.updatedAt.toISOString(),
    }));

    res.json({
      totalProjects,
      totalVideos,
      activeJobs,
      queuedJobs,
      failedJobs,
      successRate,
      totalRuntime: 0,
      projectsByStatus: projectStats,
      recentActivity,
    });
  } catch (err) {
    req.log.error(err);
    res.status(500).json({ error: "Internal server error" });
  }
});

// GET /analytics/pipeline-activity
router.get("/pipeline-activity", async (req, res) => {
  try {
    const { limit = "20" } = req.query as Record<string, string>;
    const pipelines = await db
      .select()
      .from(pipelinesTable)
      .orderBy(desc(pipelinesTable.startedAt))
      .limit(Number(limit));

    const items = pipelines.map((p) => ({
      id: p.id,
      type: "pipeline",
      message: `Pipeline ${p.status}${p.currentStage ? ` — stage: ${p.currentStage}` : ""}`,
      projectId: p.projectId,
      status: p.status,
      timestamp: p.startedAt.toISOString(),
    }));

    res.json({ items });
  } catch (err) {
    req.log.error(err);
    res.status(500).json({ error: "Internal server error" });
  }
});

// GET /analytics/stage-breakdown
router.get("/stage-breakdown", async (req, res) => {
  try {
    const stageNames = [
      "research", "script", "scene_planning", "image_generation",
      "voice_generation", "video_editing", "subtitle_generation",
      "thumbnail_generation", "seo_generation", "upload",
    ];
    const stages = stageNames.map((name) => ({
      name,
      completed: 0,
      failed: 0,
      avgDurationMs: 0,
    }));
    res.json({ stages });
  } catch (err) {
    req.log.error(err);
    res.status(500).json({ error: "Internal server error" });
  }
});

export default router;
